Add test coverage for pkg/tarsum

vdemeester opened the pull request:
Add some coverage on pkg/tarsum.

``` bash
# From
+ go test -test.timeout=30m github.com/docker/docker/pkg/tarsum
PASS
coverage: 70.0% of statements
# To
+ go test -test.timeout=30m github.com/docker/docker/pkg/tarsum
PASS
coverage: 90.3% of statements
```

And removing unused code. I was also wondering about the following :
- tarsum.go :
  NewTarSumHash could be non exported (for now)
  NewTarSumForLabel is never used, except for the tests
- fileinfosums.go:
  SortByPos is never used, except for the tests
- versionning.go:
  GetVersions is never used, expect for the tests

Signed-off-by: Vincent Demeester vincent@sbr.pm

duglin commented:
ping @vbatts

vbatts commented:
thanks. `./pkg/tarsum` will slowly get phased out, but these tests are fine. 

As for your questions, _usually_ the libraries in `./pkg/...` API are a bit fixed, because we allow for outside tooling to use them. So, while those may not be used in docker's source, we'll leave them.

vbatts commented on pkg/tarsum/builder_context_test.go:
```diff
@@ -0,0 +1,63 @@
+
+	ts.(BuilderContext).Remove("etc/sudoers")
+
+	if len(ts.GetSums()) != expected {
```
is this supposed to be a failing test?

vdemeester replied:
Hum.. no it's supposed to work :smile_cat:. I remove a *Sum* and expect one less *Sum* in the ``tarSum`` object.

vbatts replied:
Oh oh oh. I overlooked the `- 1`. Right on.
On Jun 2, 2015 02:44, "Vincent Demeester" <notifications@github.com> wrote:

> In pkg/tarsum/builder_context_test.go
> <https://github.com/docker/docker/pull/13630#discussion_r31496164>:
>
> > +	ts, err := NewTarSum(reader, false, Version0)
> > +	if err != nil {
> > +		t.Fatal(err)
> > +	}
> > +
> > +	// Read and discard bytes so that it populates sums
> > +	_, err = io.Copy(ioutil.Discard, ts)
> > +	if err != nil {
> > +		t.Errorf("failed to read from %s: %s", filename, err)
> > +	}
> > +
> > +	expected := len(ts.GetSums()) - 1
> > +
> > +	ts.(BuilderContext).Remove("etc/sudoers")
> > +
> > +	if len(ts.GetSums()) != expected {
>
> Hum.. no it's supposed to work [image: :smile_cat:]. I remove a *Sum* and
> expect one less *Sum* in the tarSum object.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13630/files#r31496164>.
>

vbatts commented:
overall, LGTM

stevvooe commented on pkg/tarsum/tarsum_test.go:
```diff
@@ -251,6 +291,33 @@ var (
+// Test all the build-in read size : buf8K, buf16K, buf32K and more
+func TestTarSumsReadSize(t *testing.T) {
+	// Test this on the four first layer
+	for i := 1; i < 6; i++ {
```
for i, layer := range testLayers {
    // ...
    nbByteToRead := (i  + 1)* 8 * 1024
    // ...
}

vdemeester replied:
@stevvooe yeah that does look way better :smile:

vdemeester replied:
Actually I just want to loop 5 times, I'm always reader the same *layer* as I'm just testing the different "read size".

stevvooe commented on pkg/tarsum/builder_context_test.go:
```diff
@@ -0,0 +1,63 @@
+
+	expected := len(ts.GetSums()) - 1
+
+	ts.(BuilderContext).Remove("etc/sudoers")
```
Probably, this isn't related to this PR, but why on earth is the builder context integrated with the `tarsum` package?

@vbatts

vbatts replied:
Caching
On Jun 1, 2015 18:28, "Stephen Day" <notifications@github.com> wrote:

> In pkg/tarsum/builder_context_test.go
> <https://github.com/docker/docker/pull/13630#discussion_r31477412>:
>
> > +		t.Fatal(err)
> > +	}
> > +	ts, err := NewTarSum(reader, false, Version0)
> > +	if err != nil {
> > +		t.Fatal(err)
> > +	}
> > +
> > +	// Read and discard bytes so that it populates sums
> > +	_, err = io.Copy(ioutil.Discard, ts)
> > +	if err != nil {
> > +		t.Errorf("failed to read from %s: %s", filename, err)
> > +	}
> > +
> > +	expected := len(ts.GetSums()) - 1
> > +
> > +	ts.(BuilderContext).Remove("etc/sudoers")
>
> Probably, this isn't related to this PR, but why on earth is the builder
> context integrated with the tarsum package?
>
> @vbatts <https://github.com/vbatts>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13630/files#r31477412>.
>

vbatts replied:
Well, for caching they needed a couple of additional features, since
certain files ought to be ignored by the .dockerignore file. There is an
extension of tarsum that allows removing certain files from the final
checksum, and this extension is called buildercontext.
On Jun 1, 2015 18:28, "Stephen Day" <notifications@github.com> wrote:

> In pkg/tarsum/builder_context_test.go
> <https://github.com/docker/docker/pull/13630#discussion_r31477412>:
>
> > +		t.Fatal(err)
> > +	}
> > +	ts, err := NewTarSum(reader, false, Version0)
> > +	if err != nil {
> > +		t.Fatal(err)
> > +	}
> > +
> > +	// Read and discard bytes so that it populates sums
> > +	_, err = io.Copy(ioutil.Discard, ts)
> > +	if err != nil {
> > +		t.Errorf("failed to read from %s: %s", filename, err)
> > +	}
> > +
> > +	expected := len(ts.GetSums()) - 1
> > +
> > +	ts.(BuilderContext).Remove("etc/sudoers")
>
> Probably, this isn't related to this PR, but why on earth is the builder
> context integrated with the tarsum package?
>
> @vbatts <https://github.com/vbatts>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13630/files#r31477412>.
>

jfrazelle commented:
LGTM thanks!

vbatts commented:
I'll wait for the couple of questions and suggestions before merging.

vdemeester committed:
Add test coverage for pkg/tarsum

And removing unused code.
- tarsum.go :
  NewTarSumHash could be non exported (for now)
  NewTarSumForLabel is never used, except for the tests
- fileinfosums.go:
  SortByPos is never used, except for the tests
- versionning.go:
  GetVersions is never used, expect for the tests

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
hum ``DockerSuite.TestCliStatsNoStreamGetCpu`` seems to fail randomly :( (failed on janky here, on windows there https://github.com/docker/docker/pull/13501)

vbatts commented:
thanks @vdemeester
